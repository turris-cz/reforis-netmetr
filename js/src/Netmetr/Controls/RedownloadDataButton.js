/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { Button } from "foris";

import { useRedownloadData } from "./hooks";

export default function RedownloadDataButton({ ws, asyncId, setAsyncId }) {
    const [onClickHandler, isLoading] = useRedownloadData(ws, asyncId, setAsyncId);

    return (
        <Button
            className="btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
            loading={isLoading}
            disabled={isLoading}
            onClick={onClickHandler}
        >
            {_("Redownload data")}
        </Button>
    );
}
