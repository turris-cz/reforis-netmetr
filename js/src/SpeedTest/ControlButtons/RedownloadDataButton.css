#redownload-data-button {
    margin-bottom: 1rem;
}
