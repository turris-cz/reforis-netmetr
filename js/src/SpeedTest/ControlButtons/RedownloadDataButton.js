/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import { Button } from "foris";
import "./RedownloadDataButton.css";

import { useRedownloadData } from "./hooks";

RedownloadDataButton.propTypes = {
    ws: PropTypes.object.isRequired,

    asyncId: PropTypes.string,
    setAsyncId: PropTypes.func.isRequired,
};

export default function RedownloadDataButton({ ws, asyncId, setAsyncId }) {
    const [onClickHandler, isLoading] = useRedownloadData(ws, asyncId, setAsyncId);

    return (
        <Button
            id="redownload-data-button"
            className="btn-outline-info btn-sm"
            loading={isLoading}
            disabled={isLoading}
            onClick={onClickHandler}
        >
            {_("Redownload data")}
        </Button>
    );
}
