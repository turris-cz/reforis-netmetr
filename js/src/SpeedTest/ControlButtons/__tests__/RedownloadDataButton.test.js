/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {render, wait, act, fireEvent} from "foris/testUtils/customTestRender";
import {WebSockets} from "foris";
import diffSnapshot from "snapshot-diff";
import mockAxios from "jest-mock-axios";

import RedownloadDataButton from '../RedownloadDataButton';

describe("<RedownloadDataButton />", () => {
    const webSockets = new WebSockets();
    const setAsyncId = jest.fn();
    let container;
    let asFragment;
    let getByText;
    let firstRender;

    beforeEach(async () => {
        ({container, asFragment, getByText} = render(
            <RedownloadDataButton ws={webSockets} setAsyncId={setAsyncId} asyncId={"test-async-id"}/>
        ));
        await wait(() => getByText("Redownload data"));
        firstRender = asFragment();
    });

    it("Should render component.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Should trigger redownload data on click.", () => {
        fireEvent.click(getByText("Redownload data"));
        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post)
            .toHaveBeenCalledWith("/reforis/netmetr/api/trigger-download-data", undefined, expect.anything());
    });

    it("Should show spinner on click.", () => {
        fireEvent.click(getByText("Redownload data"));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Should stop show spinner after WS message.", () => {
        fireEvent.click(getByText("Redownload data"));
        // Simulate receiving message from WS server.
        act(() => webSockets.dispatch({
            module: "netmetr",
            action: "download_data_finished",
            data: {
                async_id: "test-async-id",
                passed: true,
            }
        }));

        // Should be the same.
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Should not react on different asyncId in WS message.", () => {
        fireEvent.click(getByText("Redownload data"));
        // Simulate receiving message from WS server.
        act(() => webSockets.dispatch({
            module: "netmetr",
            action: "download_data_finished",
            data: {
                async_id: "other ID",
                passed: true,
            }
        }));

        // Should still show spinner.
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });
});
