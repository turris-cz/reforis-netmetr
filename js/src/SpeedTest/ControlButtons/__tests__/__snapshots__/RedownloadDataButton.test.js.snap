// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RedownloadDataButton /> Should not react on different asyncId in WS message. 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      class=\\"btn btn btn-outline-primary float-right\\"
+     disabled=\\"\\"
      id=\\"redownload-data-button\\"
      type=\\"button\\"
    >
-     Redownload data
+     <span
+       aria-hidden=\\"true\\"
+       class=\\"spinner-border spinner-border-sm\\"
+       role=\\"status\\"
+     />
+      Redownload data
    </button>
  </DocumentFragment>"
`;

exports[`<RedownloadDataButton /> Should render component. 1`] = `
<DocumentFragment>
  <button
    class="btn btn btn-outline-primary float-right"
    id="redownload-data-button"
    type="button"
  >
    Redownload data
  </button>
</DocumentFragment>
`;

exports[`<RedownloadDataButton /> Should show spinner on click. 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      class=\\"btn btn btn-outline-primary float-right\\"
+     disabled=\\"\\"
      id=\\"redownload-data-button\\"
      type=\\"button\\"
    >
-     Redownload data
+     <span
+       aria-hidden=\\"true\\"
+       class=\\"spinner-border spinner-border-sm\\"
+       role=\\"status\\"
+     />
+      Redownload data
    </button>
  </DocumentFragment>"
`;

exports[`<RedownloadDataButton /> Should stop show spinner after WS message. 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
